### Importing Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics.pairwise import cosine_similarity

### Loading the dataset

- Importing Dataset named as: netflix.csv 

In [2]:
df = pd.read_csv('netflix.csv')

- Getting the information columns that we have and checking their data-types and number instances on which columns

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


- printing First 5 instances for the reference of dataset

In [4]:
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

- We are checking for the shape which has 8807 instances per column and 12 features 

In [5]:
df.shape

(8807, 12)

- Checking for the dtypes of columns

In [6]:
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

- Checking for the null values for each columns

In [7]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

- Here we are printing rating column's unique value which would be our movie or show ratings.

In [8]:
print('Types of ratings:',df['rating'].unique())

Types of ratings: ['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']


- In this snnipet here we are searching for the total number of movies and their unique durations. 

In [9]:
print('Number of Movies:',
      df[df['type'] == 'Movie']['duration'].count(),
      '\nShow Durations:',
      df[df['type'] == 'Movie']['duration'].unique())

Number of Movies: 6128 
Show Durations: ['90 min' '91 min' '125 min' '104 min' '127 min' '67 min' '94 min'
 '161 min' '61 min' '166 min' '147 min' '103 min' '97 min' '106 min'
 '111 min' '110 min' '105 min' '96 min' '124 min' '116 min' '98 min'
 '23 min' '115 min' '122 min' '99 min' '88 min' '100 min' '102 min'
 '93 min' '95 min' '85 min' '83 min' '113 min' '13 min' '182 min' '48 min'
 '145 min' '87 min' '92 min' '80 min' '117 min' '128 min' '119 min'
 '143 min' '114 min' '118 min' '108 min' '63 min' '121 min' '142 min'
 '154 min' '120 min' '82 min' '109 min' '101 min' '86 min' '229 min'
 '76 min' '89 min' '156 min' '112 min' '107 min' '129 min' '135 min'
 '136 min' '165 min' '150 min' '133 min' '70 min' '84 min' '140 min'
 '78 min' '64 min' '59 min' '139 min' '69 min' '148 min' '189 min'
 '141 min' '130 min' '138 min' '81 min' '132 min' '123 min' '65 min'
 '68 min' '66 min' '62 min' '74 min' '131 min' '39 min' '46 min' '38 min'
 '126 min' '155 min' '159 min' '137 min' '12 min' '273 mi

- Same we are doing the shows that are presented inside our dataset

In [10]:
print('Number of Shows:',
      df[df['type'] == 'TV Show']['duration'].count(),
      '\nShow Durations:',
      df[df['type'] == 'TV Show']['duration'].unique())

Number of Shows: 2676 
Show Durations: ['2 Seasons' '1 Season' '9 Seasons' '4 Seasons' '5 Seasons' '3 Seasons'
 '6 Seasons' '7 Seasons' '10 Seasons' '8 Seasons' '17 Seasons'
 '13 Seasons' '15 Seasons' '12 Seasons' '11 Seasons']


<h4>Observations</h4>

<ul>
    <li>A lot of null values, especially for director, cast and country</li>
    <li>Rating isn't the popularity, but the suggested audience of a movie/film</li>
    <li>Listed In is essntially the genre list</li>
    <li>Duration is kept in two different metrics, time for movies and seasons for shows</li>
</ul>

# Preprocessing & Visualization

<h3>General</h3>

- Dropping Director, Cast, and Country, due to the amoung of null values and how unlikely it is that they will have a significant effect on future models

In [11]:
df = df.drop(['director', 'cast', 'country'], axis=1)

- Dropping duration column, as two different metrics are used depending on whether the item is a show or a movie. Another solution would be to split the dataset into movies and shows, but we want to be able to recommend either.

In [12]:
df = df.drop('duration', axis=1)

- Dropping date added, since it has little relevance

In [13]:
df = df.drop('date_added', axis=1)

In [14]:
df.isnull().sum()

show_id         0
type            0
title           0
release_year    0
rating          4
listed_in       0
description     0
dtype: int64

- Renaming two-word column names for ease

In [15]:
df = df.rename({'release_year':'year'}, axis=1)
df = df.rename({'listed_in':'category'}, axis=1)
df = df.rename({'show_id':'id'}, axis=1)

In [16]:
df[df['rating'].isnull()].head()

id     type                                              title  year  \
5989  s5990    Movie  13TH: A Conversation with Oprah Winfrey & Ava ...  2017   
6827  s6828  TV Show                  Gargantia on the Verdurous Planet  2013   
7312  s7313  TV Show                                       Little Lunch  2015   
7537  s7538    Movie                               My Honor Was Loyalty  2015   

     rating                              category  \
5989    NaN                                Movies   
6827    NaN  Anime Series, International TV Shows   
7312    NaN                 Kids' TV, TV Comedies   
7537    NaN                                Dramas   

                                            description  
5989  Oprah Winfrey sits down with director Ava DuVe...  
6827  After falling through a wormhole, a space-dwel...  
7312  Adopting a child's perspective, this show take...  
7537  Amid the chaos and horror of World War II, a c...

- Looking at the remaining null values, which is the four null ratings, it is low enough that we can drop those in order to simplify the computations being done

In [17]:
df = df.dropna()

In [18]:
df.shape

(8803, 7)

<h3>RS Processing</h3>

- One hot encoding the comma seperated list of genres, preparing dataframe for recommendation system

In [19]:
df_genres = df.copy()
df_genres['category'] = df_genres['category'].str.replace(' ', '')
df_enc = df_genres.join(pd.concat([df_genres['category'].str.get_dummies(sep=',')])).drop('category',axis=1)

In [20]:
pd.set_option('display.max_columns', None)
df_enc

id     type                  title  year rating  \
0        s1    Movie   Dick Johnson Is Dead  2020  PG-13   
1        s2  TV Show          Blood & Water  2021  TV-MA   
2        s3  TV Show              Ganglands  2021  TV-MA   
3        s4  TV Show  Jailbirds New Orleans  2021  TV-MA   
4        s5  TV Show           Kota Factory  2021  TV-MA   
...     ...      ...                    ...   ...    ...   
8802  s8803    Movie                 Zodiac  2007      R   
8803  s8804  TV Show            Zombie Dumb  2018  TV-Y7   
8804  s8805    Movie             Zombieland  2009      R   
8805  s8806    Movie                   Zoom  2006     PG   
8806  s8807    Movie                 Zubaan  2015  TV-14   

                                            description  Action&Adventure  \
0     As her father nears the end of his life, filmm...                 0   
1     After crossing paths at a party, a Cape Town t...                 0   
2     To protect his family from a powerful drug lor...                 0   
3     Feuds, flirtations and toilet talk go down amo...                 0   
4     In a city of coaching centers known to train I...                 0   
...                                                 ...               ...   
8802  A political cartoonist, a crime reporter and a...                 0   
8803  While living alone in a spooky town, a young g...                 0   
8804  Looking to survive in a world taken over by zo...                 0   
8805  Dragged from civilian life, a former superhero...                 0   
8806  A scrappy but poor boy worms his way into a ty...                 0   

      AnimeFeatures  AnimeSeries  BritishTVShows  Children&FamilyMovies  \
0                 0            0               0                      0   
1                 0            0               0                      0   
2                 0            0               0                      0   
3                 0            0               0                      0   
4                 0            0               0                      0   
...             ...          ...             ...                    ...   
8802              0            0               0                      0   
8803              0            0               0                      0   
8804              0            0               0                      0   
8805              0            0               0                      1   
8806              0            0               0                      0   

      Classic&CultTV  ClassicMovies  Comedies  CrimeTVShows  CultMovies  \
0                  0              0         0             0           0   
1                  0              0         0             0           0   
2                  0              0         0             1           0   
3                  0              0         0             0           0   
4                  0              0         0             0           0   
...              ...            ...       ...           ...         ...   
8802               0              0         0             0           1   
8803               0              0         0             0           0   
8804               0              0         1             0           0   
8805               0              0         1             0           0   
8806               0              0         0             0           0   

      Documentaries  Docuseries  Dramas  Faith&Spirituality  HorrorMovies  \
0                 1           0       0                   0             0   
1                 0           0       0                   0             0   
2                 0           0       0                   0             0   
3                 0           1       0                   0             0   
4                 0           0       0                   0             0   
...             ...         ...     ...                 ...           ...   
8802              0   

<h4>Renaming columns</h4>

- Noticing a lot of redundancy between categories, especially to split movies and shows. I am going to essentially merge the two

In [21]:
def getFirstRow(df,feature):
    return df[df[feature] == 1].head(1)

In [22]:
def getRow(df,id):
    return df.loc[df_enc['id'] == id]

In [23]:
df_enc.columns

Index(['id', 'type', 'title', 'year', 'rating', 'description',
       'Action&Adventure', 'AnimeFeatures', 'AnimeSeries', 'BritishTVShows',
       'Children&FamilyMovies', 'Classic&CultTV', 'ClassicMovies', 'Comedies',
       'CrimeTVShows', 'CultMovies', 'Documentaries', 'Docuseries', 'Dramas',
       'Faith&Spirituality', 'HorrorMovies', 'IndependentMovies',
       'InternationalMovies', 'InternationalTVShows', 'Kids'TV',
       'KoreanTVShows', 'LGBTQMovies', 'Movies', 'Music&Musicals', 'RealityTV',
       'RomanticMovies', 'RomanticTVShows', 'Sci-Fi&Fantasy',
       'Science&NatureTV', 'Spanish-LanguageTVShows', 'SportsMovies',
       'Stand-UpComedy', 'Stand-UpComedy&TalkShows', 'TVAction&Adventure',
       'TVComedies', 'TVDramas', 'TVHorror', 'TVMysteries', 'TVSci-Fi&Fantasy',
       'TVShows', 'TVThrillers', 'TeenTVShows', 'Thrillers'],
      dtype='object')

- Getting thefirst row AnimeFeatures and from here I will be assinging the repsective column to 1 for the encoding puprose

In [24]:
getFirstRow(df_enc,'AnimeFeatures')

id   type                                              title  year  \
51  s52  Movie  InuYasha the Movie 2: The Castle Beyond the Lo...  2002   

   rating                                        description  \
51  TV-14  With their biggest foe seemingly defeated, Inu...   

    Action&Adventure  AnimeFeatures  AnimeSeries  BritishTVShows  \
51                 1              1            0               0   

    Children&FamilyMovies  Classic&CultTV  ClassicMovies  Comedies  \
51                      0               0              0         0   

    CrimeTVShows  CultMovies  Documentaries  Docuseries  Dramas  \
51             0           0              0           0       0   

    Faith&Spirituality  HorrorMovies  IndependentMovies  InternationalMovies  \
51                   0             0                  0                    1   

    InternationalTVShows  Kids'TV  KoreanTVShows  LGBTQMovies  Movies  \
51                     0        0              0            0       0   

    Music&Musicals  RealityTV  RomanticMovies  RomanticTVShows  \
51               0          0               0                0   

    Sci-Fi&Fantasy  Science&NatureTV  Spanish-LanguageTVShows  SportsMovies  \
51               0                 0                        0             0   

    Stand-UpComedy  Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  \
51               0                         0                   0           0   

    TVDramas  TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  \
51         0         0            0                 0        0            0   

    TeenTVShows  Thrillers  
51            0          0

- Here we are renaming the AnimeFeatures to Anime and placing it where AnimeSeries to 1. I will be repeating this thing for every categories or movie and shows

In [25]:
df_enc = df_enc.rename({'AnimeFeatures':'Anime'}, axis=1)
df_enc.loc[df_enc['AnimeSeries'] == 1, 'Anime'] = 1
df_enc = df_enc.drop('AnimeSeries', axis=1)

In [26]:
getRow(df_enc,'s77')

id     type            title  year rating  \
76  s77  TV Show  Yowamushi Pedal  2013  TV-14   

                                          description  Action&Adventure  \
76  A timid, anime-loving teen gets drawn into a s...                 0   

    Anime  BritishTVShows  Children&FamilyMovies  Classic&CultTV  \
76      1               0                      0               0   

    ClassicMovies  Comedies  CrimeTVShows  CultMovies  Documentaries  \
76              0         0             0           0              0   

    Docuseries  Dramas  Faith&Spirituality  HorrorMovies  IndependentMovies  \
76           0       0                   0             0                  0   

    InternationalMovies  InternationalTVShows  Kids'TV  KoreanTVShows  \
76                    0                     1        0              0   

    LGBTQMovies  Movies  Music&Musicals  RealityTV  RomanticMovies  \
76            0       0               0          0               0   

    RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  \
76                0               0                 0   

    Spanish-LanguageTVShows  SportsMovies  Stand-UpComedy  \
76                        0             0               0   

    Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  TVDramas  \
76                         0                   0           0         0   

    TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  \
76         0            0                 0        0            0   

    TeenTVShows  Thrillers  
76            0          0

<h5>Classical and Cult</h5>

In [27]:
getFirstRow(df_enc,'Classic&CultTV')

id     type             title  year rating  \
390  s391  TV Show  The Walking Dead  2019  TV-MA   

                                           description  Action&Adventure  \
390  In the wake of a zombie apocalypse, survivors ...                 0   

     Anime  BritishTVShows  Children&FamilyMovies  Classic&CultTV  \
390      0               0                      0               1   

     ClassicMovies  Comedies  CrimeTVShows  CultMovies  Documentaries  \
390              0         0             0           0              0   

     Docuseries  Dramas  Faith&Spirituality  HorrorMovies  IndependentMovies  \
390           0       0                   0             0                  0   

     InternationalMovies  InternationalTVShows  Kids'TV  KoreanTVShows  \
390                    0                     0        0              0   

     LGBTQMovies  Movies  Music&Musicals  RealityTV  RomanticMovies  \
390            0       0               0          0               0   

     RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  \
390                0               0                 0   

     Spanish-LanguageTVShows  SportsMovies  Stand-UpComedy  \
390                        0             0               0   

     Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  TVDramas  \
390                         0                   1           0         1   

     TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  \
390         0            0                 0        0            0   

     TeenTVShows  Thrillers  
390            0          0

- Spilling Classic and Cult into two, assigning Classic&Cult to both, then comboning classicTV and classicMovies

In [28]:

df_enc = df_enc.rename({'Classic&CultTV':'Classic'}, axis=1)

df_enc = df_enc.rename({'CultMovies':'Cult'}, axis=1)
df_enc.loc[df_enc['Classic'] == 1, 'Cult'] = 1

df_enc.loc[df_enc['ClassicMovies'] == 1, 'Classic'] = 1
df_enc = df_enc.drop('ClassicMovies', axis=1)

In [29]:
getRow(df_enc,'s391')

id     type             title  year rating  \
390  s391  TV Show  The Walking Dead  2019  TV-MA   

                                           description  Action&Adventure  \
390  In the wake of a zombie apocalypse, survivors ...                 0   

     Anime  BritishTVShows  Children&FamilyMovies  Classic  Comedies  \
390      0               0                      0        1         0   

     CrimeTVShows  Cult  Documentaries  Docuseries  Dramas  \
390             0     1              0           0       0   

     Faith&Spirituality  HorrorMovies  IndependentMovies  InternationalMovies  \
390                   0             0                  0                    0   

     InternationalTVShows  Kids'TV  KoreanTVShows  LGBTQMovies  Movies  \
390                     0        0              0            0       0   

     Music&Musicals  RealityTV  RomanticMovies  RomanticTVShows  \
390               0          0               0                0   

     Sci-Fi&Fantasy  Science&NatureTV  Spanish-LanguageTVShows  SportsMovies  \
390               0                 0                        0             0   

     Stand-UpComedy  Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  \
390               0                         0                   1           0   

     TVDramas  TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  \
390         1         0            0                 0        0            0   

     TeenTVShows  Thrillers  
390            0          0

<h5>Documentaries</h5>

In [30]:
getFirstRow(df_enc,'Docuseries')

id     type                  title  year rating  \
3  s4  TV Show  Jailbirds New Orleans  2021  TV-MA   

                                         description  Action&Adventure  Anime  \
3  Feuds, flirtations and toilet talk go down amo...                 0      0   

   BritishTVShows  Children&FamilyMovies  Classic  Comedies  CrimeTVShows  \
3               0                      0        0         0             0   

   Cult  Documentaries  Docuseries  Dramas  Faith&Spirituality  HorrorMovies  \
3     0              0           1       0                   0             0   

   IndependentMovies  InternationalMovies  InternationalTVShows  Kids'TV  \
3                  0                    0                     0        0   

   KoreanTVShows  LGBTQMovies  Movies  Music&Musicals  RealityTV  \
3              0            0       0               0          1   

   RomanticMovies  RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  \
3               0                0               0                 0   

   Spanish-LanguageTVShows  SportsMovies  Stand-UpComedy  \
3                        0             0               0   

   Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  TVDramas  \
3                         0                   0           0         0   

   TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  TeenTVShows  \
3         0            0                 0        0            0            0   

   Thrillers  
3          0

In [31]:
df_enc.loc[df_enc['Docuseries'] == 1, 'Documentaries'] = 1
df_enc = df_enc.drop('Docuseries', axis=1)

In [32]:
getRow(df_enc,'s4')

id     type                  title  year rating  \
3  s4  TV Show  Jailbirds New Orleans  2021  TV-MA   

                                         description  Action&Adventure  Anime  \
3  Feuds, flirtations and toilet talk go down amo...                 0      0   

   BritishTVShows  Children&FamilyMovies  Classic  Comedies  CrimeTVShows  \
3               0                      0        0         0             0   

   Cult  Documentaries  Dramas  Faith&Spirituality  HorrorMovies  \
3     0              1       0                   0             0   

   IndependentMovies  InternationalMovies  InternationalTVShows  Kids'TV  \
3                  0                    0                     0        0   

   KoreanTVShows  LGBTQMovies  Movies  Music&Musicals  RealityTV  \
3              0            0       0               0          1   

   RomanticMovies  RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  \
3               0                0               0                 0   

   Spanish-LanguageTVShows  SportsMovies  Stand-UpComedy  \
3                        0             0               0   

   Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  TVDramas  \
3                         0                   0           0         0   

   TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  TeenTVShows  \
3         0            0                 0        0            0            0   

   Thrillers  
3          0

<h5>Family/Kids</h5>

In [33]:
getFirstRow(df_enc,'Children&FamilyMovies')

id   type                             title  year rating  \
6  s7  Movie  My Little Pony: A New Generation  2021     PG   

                                         description  Action&Adventure  Anime  \
6  Equestria's divided. But a bright-eyed hero be...                 0      0   

   BritishTVShows  Children&FamilyMovies  Classic  Comedies  CrimeTVShows  \
6               0                      1        0         0             0   

   Cult  Documentaries  Dramas  Faith&Spirituality  HorrorMovies  \
6     0              0       0                   0             0   

   IndependentMovies  InternationalMovies  InternationalTVShows  Kids'TV  \
6                  0                    0                     0        0   

   KoreanTVShows  LGBTQMovies  Movies  Music&Musicals  RealityTV  \
6              0            0       0               0          0   

   RomanticMovies  RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  \
6               0                0               0                 0   

   Spanish-LanguageTVShows  SportsMovies  Stand-UpComedy  \
6                        0             0               0   

   Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  TVDramas  \
6                         0                   0           0         0   

   TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  TeenTVShows  \
6         0            0                 0        0            0            0   

   Thrillers  
6          0

In [34]:
# Essentially, splitting Family and Children/Kids categories and adding Children&Family to both
df_enc = df_enc.rename({'Children&FamilyMovies':'Family'}, axis=1)
df_enc = df_enc.rename({"Kids'TV":'Kids'}, axis=1)

df_enc.loc[df_enc['Family'] == 1, 'Kids'] = 1

In [35]:
getRow(df_enc,'s7')

id   type                             title  year rating  \
6  s7  Movie  My Little Pony: A New Generation  2021     PG   

                                         description  Action&Adventure  Anime  \
6  Equestria's divided. But a bright-eyed hero be...                 0      0   

   BritishTVShows  Family  Classic  Comedies  CrimeTVShows  Cult  \
6               0       1        0         0             0     0   

   Documentaries  Dramas  Faith&Spirituality  HorrorMovies  IndependentMovies  \
6              0       0                   0             0                  0   

   InternationalMovies  InternationalTVShows  Kids  KoreanTVShows  \
6                    0                     0     1              0   

   LGBTQMovies  Movies  Music&Musicals  RealityTV  RomanticMovies  \
6            0       0               0          0               0   

   RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  Spanish-LanguageTVShows  \
6                0               0                 0                        0   

   SportsMovies  Stand-UpComedy  Stand-UpComedy&TalkShows  TVAction&Adventure  \
6             0               0                         0                   0   

   TVComedies  TVDramas  TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  \
6           0         0         0            0                 0        0   

   TVThrillers  TeenTVShows  Thrillers  
6            0            0          0

<h5>International</h5>

In [36]:
getFirstRow(df_enc,'InternationalMovies')

id   type    title  year rating  \
7  s8  Movie  Sankofa  1993  TV-MA   

                                         description  Action&Adventure  Anime  \
7  On a photo shoot in Ghana, an American model s...                 0      0   

   BritishTVShows  Family  Classic  Comedies  CrimeTVShows  Cult  \
7               0       0        0         0             0     0   

   Documentaries  Dramas  Faith&Spirituality  HorrorMovies  IndependentMovies  \
7              0       1                   0             0                  1   

   InternationalMovies  InternationalTVShows  Kids  KoreanTVShows  \
7                    1                     0     0              0   

   LGBTQMovies  Movies  Music&Musicals  RealityTV  RomanticMovies  \
7            0       0               0          0               0   

   RomanticTVShows  Sci-Fi&Fantasy  Science&NatureTV  Spanish-LanguageTVShows  \
7                0               0                 0                        0   

   SportsMovies  Stand-UpComedy  Stand-UpComedy&TalkShows  TVAction&Adventure  \
7             0               0                         0                   0   

   TVComedies  TVDramas  TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  \
7           0         0         0            0                 0        0   

   TVThrillers  TeenTVShows  Thrillers  
7            0            0          0

In [37]:
df_enc = df_enc.rename({'InternationalMovies':'International'}, axis=1)
df_enc.loc[df_enc['InternationalTVShows'] == 1, 'International'] = 1
df_enc = df_enc.drop('InternationalTVShows', axis=1)

In [38]:
getRow(df_enc,'s8')

id   type    title  year rating  \
7  s8  Movie  Sankofa  1993  TV-MA   

                                         description  Action&Adventure  Anime  \
7  On a photo shoot in Ghana, an American model s...                 0      0   

   BritishTVShows  Family  Classic  Comedies  CrimeTVShows  Cult  \
7               0       0        0         0             0     0   

   Documentaries  Dramas  Faith&Spirituality  HorrorMovies  IndependentMovies  \
7              0       1                   0             0                  1   

   International  Kids  KoreanTVShows  LGBTQMovies  Movies  Music&Musicals  \
7              1     0              0            0       0               0   

   RealityTV  RomanticMovies  RomanticTVShows  Sci-Fi&Fantasy  \
7          0               0                0               0   

   Science&NatureTV  Spanish-LanguageTVShows  SportsMovies  Stand-UpComedy  \
7                 0                        0             0               0   

   Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  TVDramas  \
7                         0                   0           0         0   

   TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  TeenTVShows  \
7         0            0                 0        0            0            0   

   Thrillers  
7          0

<h5>Comedy And Talk Shows</h5>

In [39]:
getFirstRow(df_enc,'Stand-UpComedy')

id   type                        title  year rating  \
278  s279  Movie  Lokillo: Nothing's the Same  2021  TV-MA   

                                           description  Action&Adventure  \
278  Through songs and puns, comedian Lokillo Flore...                 0   

     Anime  BritishTVShows  Family  Classic  Comedies  CrimeTVShows  Cult  \
278      0               0       0        0         0             0     0   

     Documentaries  Dramas  Faith&Spirituality  HorrorMovies  \
278              0       0                   0             0   

     IndependentMovies  International  Kids  KoreanTVShows  LGBTQMovies  \
278                  0              0     0              0            0   

     Movies  Music&Musicals  RealityTV  RomanticMovies  RomanticTVShows  \
278       0               0          0               0                0   

     Sci-Fi&Fantasy  Science&NatureTV  Spanish-LanguageTVShows  SportsMovies  \
278               0                 0                        0             0   

     Stand-UpComedy  Stand-UpComedy&TalkShows  TVAction&Adventure  TVComedies  \
278               1                         0                   0           0   

     TVDramas  TVHorror  TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  \
278         0         0            0                 0        0            0   

     TeenTVShows  Thrillers  
278            0          0

- Puts stand-up comedy into talk shows and comedies before dropping the column

In [40]:

df_enc = df_enc.rename({'Comedies':'Comedy'}, axis=1)

df_enc = df_enc.rename({'Stand-UpComedy&TalkShows':'TalkShow'}, axis=1)
df_enc.loc[df_enc['Stand-UpComedy'] == 1, 'TalkShow'] = 1

df_enc.loc[df_enc['Stand-UpComedy'] == 1, 'Comedy'] = 1
df_enc = df_enc.drop('Stand-UpComedy', axis=1)

In [41]:
getRow(df_enc,'s279')

id   type                        title  year rating  \
278  s279  Movie  Lokillo: Nothing's the Same  2021  TV-MA   

                                           description  Action&Adventure  \
278  Through songs and puns, comedian Lokillo Flore...                 0   

     Anime  BritishTVShows  Family  Classic  Comedy  CrimeTVShows  Cult  \
278      0               0       0        0       1             0     0   

     Documentaries  Dramas  Faith&Spirituality  HorrorMovies  \
278              0       0                   0             0   

     IndependentMovies  International  Kids  KoreanTVShows  LGBTQMovies  \
278                  0              0     0              0            0   

     Movies  Music&Musicals  RealityTV  RomanticMovies  RomanticTVShows  \
278       0               0          0               0                0   

     Sci-Fi&Fantasy  Science&NatureTV  Spanish-LanguageTVShows  SportsMovies  \
278               0                 0                        0             0   

     TalkShow  TVAction&Adventure  TVComedies  TVDramas  TVHorror  \
278         1                   0           0         0         0   

     TVMysteries  TVSci-Fi&Fantasy  TVShows  TVThrillers  TeenTVShows  \
278            0                 0        0            0            0   

     Thrillers  
278          0

<h5>Misc. Renaming</h5>
- Here I am renaming and assigning 1 to repsective column for the encoding purpose

In [42]:
df_enc = df_enc.rename({'BritishTVShows':'British'}, axis=1)
df_enc = df_enc.rename({'CrimeTVShows':'Crime'}, axis=1)
df_enc = df_enc.rename({'Faith&Spirituality':'Faith'}, axis=1)
df_enc = df_enc.rename({'IndependentMovies':'Independant'}, axis=1)
df_enc = df_enc.rename({'KoreanTVShows':'Korean'}, axis=1)
df_enc = df_enc.rename({'LGBTQMovies':'LGBTQ'}, axis=1)
df_enc = df_enc.rename({'Music&Musicals':'Musical'}, axis=1)
df_enc = df_enc.rename({'RealityTV':'Reality'}, axis=1)
df_enc = df_enc.rename({'Science&NatureTV':'Science'}, axis=1)
df_enc = df_enc.rename({'Spanish-LanguageTVShows':'Spanish'}, axis=1)
df_enc = df_enc.rename({'SportsMovies':'Sports'}, axis=1)
df_enc = df_enc.rename({'TVMysteries':'Mystery'}, axis=1)
df_enc = df_enc.rename({'TeenTVShows':'Teen'}, axis=1)

In [43]:
df_enc = df_enc.drop('Movies', axis=1)
df_enc = df_enc.drop('TVShows', axis=1)

In [44]:
df_enc.loc[df_enc['TVDramas'] == 1, 'Dramas'] = 1
df_enc = df_enc.drop('TVDramas', axis=1)

In [45]:
df_enc = df_enc.rename({'Action&Adventure':'Action'}, axis=1)
df_enc.loc[df_enc['TVAction&Adventure'] == 1, 'Action'] = 1
df_enc = df_enc.drop('TVAction&Adventure', axis=1)

In [46]:
df_enc = df_enc.rename({'RomanticMovies':'Romance'}, axis=1)
df_enc.loc[df_enc['RomanticTVShows'] == 1, 'Romance'] = 1
df_enc = df_enc.drop('RomanticTVShows', axis=1)

In [47]:
df_enc = df_enc.rename({'HorrorMovies':'Horror'}, axis=1)
df_enc.loc[df_enc['TVHorror'] == 1, 'Horror'] = 1
df_enc = df_enc.drop('TVHorror', axis=1)

In [48]:
df_enc = df_enc.rename({'Sci-Fi&Fantasy':'SciFiFan'}, axis=1)
df_enc.loc[df_enc['TVSci-Fi&Fantasy'] == 1, 'SciFiFan'] = 1
df_enc = df_enc.drop('TVSci-Fi&Fantasy', axis=1)

In [49]:
df_enc.loc[df_enc['TVThrillers'] == 1, 'Thrillers'] = 1
df_enc = df_enc.drop('TVThrillers', axis=1)


- Extract more features from the description column
- For simplicity, we'll use the length of the description as a feature

In [50]:
df['description_length'] = df['description'].apply(lambda x: len(str(x)))
df['description_length'].head()

0    152
1    147
2    146
3    147
4    152
Name: description_length, dtype: int64

<h3>Visualizations</h3>